src/core/Tree/TreeNode.tsx

```diff
@@ -0,0 +1,270 @@
+  /**
+   * Content shown after `TreeNode`.
+   */
+  children?: React.ReactNode;
+} & Omit<CommonProps, 'id'>;
+
+/**
+ * `TreeNode` component to display node content within a `Tree`.
+ * Must be used inside `Tree` component to correctly set node `depth` and `subNodes`.
+ * @example
+  <TreeNode
+    nodeId={props.nodeId}
+    label={props.node.label}
+    sublabel={props.node.sublabel}
+    onExpanded={onExpanded}
+    onSelected={onSelectedNodeChange}
+    isDisabled={props.isDisabled}
+    isExpanded={props.isExpanded}
+    isSelected={props.isSelected}
+    checkbox={
+      <Checkbox variant='eyeball' disabled={props.isDisabled} />
+    }
+    icon={<SvgPlaceholder />}
+  />
+ */
+export const TreeNode = (props: TreeNodeProps) => {
+  const {
+    nodeId,
+    label,
+    sublabel,
+    children,
+    style,
+    className,
+    icon,
+    hasSubNodes = false,
+    isDisabled = false,
+    isExpanded = false,
+    isSelected = false,
+    onSelected,
+    onExpanded,
+    checkbox,
+    expander,
+    ...rest
+  } = props;
+  useTheme();
+
+  const context = React.useContext(TreeContext);
+  const nodeDepth = context?.nodeDepth ?? 0;
+  const subNodeIds = context?.subNodeIds ?? [];
+  const parentNodeId = context?.parentNodeId;
+  const groupSize = context?.groupSize ?? 0;
+  const indexInGroup = context?.indexInGroup ?? 0;
+
+  const [isFocused, setIsFocused] = React.useState(false);
+  const nodeRef = React.useRef<HTMLLIElement>(null);
+
+  const styleDepth = React.useMemo(
+    () =>
+      getWindow()?.CSS?.supports?.(`--level: ${nodeDepth}`)
+        ? { '--level': nodeDepth, ...style }
+        : { marginLeft: nodeDepth ? nodeDepth * 28 : 0, ...style },
+    [nodeDepth, style],
+  );
+
+  const onKeyDown = (event: React.KeyboardEvent<HTMLLIElement>) => {
+    switch (event.key) {
+      case 'ArrowLeft':
+        event.preventDefault();
+        if (isExpanded) {
+          onExpanded(nodeId, false);
+        } else if (parentNodeId) {
+          const parentNode = nodeRef.current?.ownerDocument.querySelector(
+            `#${parentNodeId}`,
+          ) as HTMLElement;
+          parentNode?.focus();
+        }
+        break;
+      case 'ArrowRight':
+        event.preventDefault();
+        if (!isExpanded && hasSubNodes) {
+          onExpanded(nodeId, true);
+        } else if (subNodeIds.length > 0) {
+          const subNodes = Array.from(
+            nodeRef.current?.ownerDocument.querySelectorAll(
+              `${subNodeIds.map((id) => `#${id}`).join(',')}`,
+            ) || [],
+          ).filter(
+            (el) => el.getAttribute('aria-disabled') !== 'true',
+          ) as HTMLElement[];
+          subNodes[0]?.focus();
+        }
+        break;
+      case ' ':
+      case 'Spacebar':
+      case 'Enter':
+        // Ignore if it is called on the element inside, e.g. checkbox or expander
+        if (event.target !== nodeRef.current) {
+          break;
+        }
+        event.preventDefault();
```

Comment:
Yeah, it is the issue with the checkbox, I hope your PR will fix it, because it is working with expander button.
As you said it yourself there is no standard for navigating inside node, so I don't want to write this ugly handler and check whether it is first or last item and do actions accordingly, so I made actionable items tabbable and that really simplifies everything.

Reply:
By that logic, everything should be tabbable. Just because the code is simpler doesn't mean it's the best experience. https://twitter.com/annaecook/status/1494340020144001027

Reply:
I didn't go with the easiest path as it was handling arrow keys navigation but nvm... Changed, check it, I can't say it is better than it was previously.

Reply:
It is much better now!

I think there is one small issue: currently if a node is collapsed and I press right arrow, then it traverses through the checkbox, expander, etc. I think it should first expand the node, and only after that traverse through the contents.

See this example: https://www.w3.org/TR/wai-aria-practices-1.2/examples/treegrid/treegrid-1.html

Reply:
This example is great! Now I prefer this keyboard navigation over my old one 😀 Btw, changed.

Reply:
There seems to be a bug 🤔 After reaching the rightmost element in the node, pressing Right arrow goes into the next node's checkbox and stops.

Reply:
But it's really confusing when combined with the new behavior of allowing traversing contents. See gif.

In this gif, I'm only using left and right keys. I start on Node 2, press Right to focus checkbox and then expander, and then again to focus Node 2-0, and then again to focus checkbox in Node 2-0. But then it stops! I can press Left to go back but I can't press Right to go to Node 2-1.
![tree](https://user-images.githubusercontent.com/9084735/155413632-4ec66f4d-dc77-4c86-a913-a04194b3fb92.gif)

The suggestions from your screenshot were more relevant before we made this whole set of changes for allowing content traversal. What we want is something more like this:
![image](https://user-images.githubusercontent.com/9084735/155414176-648b472a-5f84-4e57-9ed7-5580636ddfb0.png)
https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-25

Reply:
Fixed.